tests/lib/thrift/ThriftTest/src/client.cpp

```diff
@@ -0,0 +1,509 @@
+    request_set.insert(i);
+  }
+
+  set<int32_t> response_set;
+  testClient.testSet(request_set, response_set);
+
+  test_equal_size(request_set, response_set);
+  test_identity_forEach(request_set.cbegin(), request_set.cend(), response_set.cbegin());
+}
+
+static void list_test(void) {
+  ThriftTestClient& testClient = *(context.client);
+  try {
+    vector<int32_t> response_list;
+    testClient.testList(response_list, vector<int32_t>());
+
+    zassert_true(response_list.empty(), "Unexpected list size: %llu", response_list.size());
+  } catch (TTransportException&) {
+    throw;
+  } catch (exception& ex) {
+    zassert_exception("%s", ex.what());
+  }
+
+  try {
+    vector<int32_t> request_list;
+    for (int32_t i = -2; i < 3; ++i) {
+      request_list.push_back(i);
+    }
+
+    vector<int32_t> response_list;
+    testClient.testList(response_list, request_list);
+
+    test_equal_size(request_list, response_list);
+    test_identity_forEach(request_list.cbegin(), request_list.cend(), response_list.cbegin());
+  } catch (TTransportException&) {
+    throw;
+  } catch (exception& ex) {
+    zassert_exception("%s", ex.what());
+  }
+}
+
+static void enum_test(void) {
+  ThriftTestClient& testClient = *(context.client);
+  Numberz::type response = testClient.testEnum(Numberz::ONE);
+  zassert_equal(response, Numberz::ONE, NULL);
+
+  response = testClient.testEnum(Numberz::TWO);
+  zassert_equal(response, Numberz::TWO, NULL);
+
+  response = testClient.testEnum(Numberz::EIGHT);
+  zassert_equal(response, Numberz::EIGHT, NULL);
+}
+
+static void typedef_test(void) {
+  ThriftTestClient& testClient = *(context.client);
+  UserId uid = testClient.testTypedef(309858235082523LL);
+  zassert_equal(uid, 309858235082523LL, "Unexpected uid: %llu", uid);
+}
+
+static void nested_map_test(void) {
+  ThriftTestClient& testClient = *(context.client);
+  map<int32_t, map<int32_t, int32_t>> mm;
+  testClient.testMapMap(mm, 1);
+
+  zassert_equal(mm.size(), 2, NULL);
+  zassert_equal(mm[-4][-4], -4, NULL);
+  zassert_equal(mm[-4][-3], -3, NULL);
+  zassert_equal(mm[-4][-2], -2, NULL);
+  zassert_equal(mm[-4][-1], -1, NULL);
+  zassert_equal(mm[4][4], 4, NULL);
+  zassert_equal(mm[4][3], 3, NULL);
+  zassert_equal(mm[4][2], 2, NULL);
+  zassert_equal(mm[4][1], 1, NULL);
+}
+
+static void insanity_test(void) {
+  ThriftTestClient& testClient = *(context.client);
+  Insanity insane;
+  insane.userMap.insert(make_pair(Numberz::FIVE, 5));
+  insane.userMap.insert(make_pair(Numberz::EIGHT, 8));
+  Xtruct truck;
+  truck.string_thing = "Goodbye4";
+  truck.byte_thing = 4;
+  truck.i32_thing = 4;
+  truck.i64_thing = 4;
+  Xtruct truck2;
+  truck2.string_thing = "Hello2";
+  truck2.byte_thing = 2;
+  truck2.i32_thing = 2;
+  truck2.i64_thing = 2;
+  insane.xtructs.push_back(truck);
+  insane.xtructs.push_back(truck2);
+
+  map<UserId, map<Numberz::type, Insanity>> whoa;
+  testClient.testInsanity(whoa, insane);
+
+  zassert_equal(whoa.size(), 2, NULL);
```

Comment:
As I mentioned on discord, rather than testing every element of containers, structures, etc, simply overload `operator==` for the types you are interested in comparing and then use `zassert_true(foo == bar, "")`. If ever the test fails, a debugger can be attached to examine details.

E.g. 
```cpp
template <class T>
bool operator==(const vector<T>& a, const vector<T>& b)
{
  if (a.size() != b.size()) {
    return false;
  }
  
  for(auto i = 0, n = a.size(); i < n; ++i) {
    if (a[i] != b[i]) {
      return false;
    }
  }
  
  return true;
}
```

Reply:
In fact, I want to throw the entire "insanity" test away. As nested structures and maps were covered in other tests, I don't find this test particularly useful. Plus the original code of this test from the upstream is a mess.

Reply:
Also, thrift has generated `operator ==` for structures defined in IDL in `*_types.h`. Those comparators do deep comparison, thus there is no need to overload them manually. 

Example of generated `operator ==` in `ThriftTest_types.h`:
```
  bool operator == (const Xtruct3 & rhs) const
  {
    if (!(string_thing == rhs.string_thing))
      return false;
    if (!(changed == rhs.changed))
      return false;
    if (!(i32_thing == rhs.i32_thing))
      return false;
    if (!(i64_thing == rhs.i64_thing))
      return false;
    return true;
  }
```